src/typescript/frontend/src/hooks/use-is-banned.ts

```diff
@@ -0,0 +1,12 @@
+import { useState } from "react";
+import { isBanned } from "utils/geolocation";
+
+const useIsBanned = () => {
+  const [banned, setBanned] = useState(true);
+
+  isBanned().then((res) => setBanned(res));
+
+  return banned;
+};
```

Comment:
Unfortunately this will call the function dozens of times- you could use `useRef` here to make the value stable across re-renders, but even then it'd still call it every time you use the hook in a different component.

If you had to do it yourself, since it's an app-wide global state check, something like this should really be in its own context provider or in a zustand state store to ensure there's a single state for the promise. Plus, you probably also want to implement locking across multiple components as well

But since this is just a fetch call this is what `@tanstack/react-query` is for really, because it has its own app-wide global cache and key mechanism already 😄 

Or you can do it server side and cache the result. React/nextjs purportedly say multiple server-side async fetches are stable across several components, so you theoretically wouldn't have to care about state

Reply:
This is not a problem.

Using the `LazyPromise` object, the calculations only happen once.

To make sure of what I'm saying, I added a `console.log` statement in the `getCountry` function, and I only see it printed once.

Yes, `isBanned` might be called multiple times, but here is the code of that function:

```
if (!GEOBLOCKING_ENABLED) return false;
return !isAllowedCountry(await country.get());
```

First, it's checking a constant. Then, it's calling `isAllowedCountry` (which is just a comparison) on `await country.get()`. `country` here is the `LazyPromise`, which basically ensures that the calculation only happens once, and all subsequent requests are served the computed value obtained on the first calculation.

It works. If you're ok with this, I wouldn't change it.

Reply:
I think it would mostly *work* as long as it stays exactly the way it is, but if there are any changes to how we manage geoblocking it will almost certainly need to be changed

It's also just generally not a great practice to put something in the function body if the intention isn't to call the function each render cycle. You could achieve the same exact effect you have now with a `useEffect` or even a `useEffect` that stores a boolean ref (to be stable across component mount/unmount).

But in that case again you're asusming the state would never change, and it might, if for example, the fetch call fails and we need to re-fetch